### PR TITLE
Allow restricting tag generation to files listed by custom commands

### PR DESCRIPTION
**Pull request by smkent**
This adds a new setting, `g:gutentags_file_list_command`, which specifies command(s) to use to list files for which tags should be generated, instead of recursively examining all files within the project root. This implements the feature requested in https://github.com/ludovicchabant/vim-gutentags/issues/90.

This is useful in projects using source control to restrict tag generation to only files tracked in the repository.

This setting is conceptually similar to [CtrlP](https://github.com/ctrlpvim/ctrlp.vim)'s [`g:ctrlp_user_command`](https://github.com/ctrlpvim/ctrlp.vim/blob/master/doc/ctrlp.txt#L302-L368) option and accepts similar types of values. For example, to configure vim-gutentags with different file list commands for source control systems, this option can be set as follows:

```vim
let g:gutentags_file_list_command = {
    \ 'types': {
        \ 1: ['.git', 'git ls-files'],
        \ 2: ['.hg', 'hg locate'],
        \ },
    \ }
```

See the changes to `doc/gutentags.txt` for full details on what values are accepted by `g:gutentags_file_list_command`.

**Notes:**
* I primarily use `ctags`. While this change also supports specifying a file list command with `cscope`, I haven't done a large amount of testing. `cscope` accepts a list of files to examine using `cscope -i -`, similar to `ctags -L -`.
* I am not able to fully test the changes to the Windows scripts, although I did test the basic syntax and changes to conditionals. While I am hopeful these changes will work properly on Windows, I do not have a full setup available to test in vim on Windows.

### CODE REVIEW

**Comment by ludovicchabant** [comment]
Looks good, mostly, thanks! I'm on vacation though so I only have a Mac with me. This will have to wait until I'm back to I can test the Windows version on my other computer.

I'd like to see the way the new option works to be changed though:

* The `List` option adds a bit of complexity to the code and documentation for no real reason -- Gutentags could just support the `String` and `Dictionary` forms and that would be enough I think (even if the user only wants a file-listing command for one type of repository, writing it as a `Dictionary` isn't much of a hassle, and sets you up for the future).

* In the `Dictionary` form, I would make the main key name be `roots` or `markers` instead of `type` -- `type` to me implies a match on the project's type (_i.e._ "python", or "ruby" or whatever).

* Does the `Dictionary` form require any sort of ordering? I imagine not, so I would prefer to see something like this instead:

```
let g:gutentags_file_list_command = {
    \"markers": {
    \    ".git": "git ls-files",
    \    ".hg": "hg files"}
    \}
```

Basically, make the root marker the key in the mapping.

**Comment by smkent** [comment]
Thanks for taking a look, I'm glad you like the feature! It definitely makes Gutentags easily usable for me in several of my projects.

I'll make the suggested changes to the values accepted by `g:gutentags_file_list_command`.

The vim documentation says `Dictionary` keys are "stored without a specific ordering," but vim does seem to iterate over them in alphabetical order. I think that's why `ctrlp_user_command`'s `Dictionary` format is the way it is -- it gives the user a way to order the various marker types by preference. For Gutentags, I can see how the simpler format would be preferable, since there should rarely (or never) be multiple repository markers in the same project root. I could even reduce the `Dictionary` format to just marker <-> command pairs if you like:

```vim
let g:gutentags_file_list_command = {
    \ '.git': 'git ls-files',
    \ '.hg': 'hg files',
    \ }
```

I've also found that `gutentags_file_list_command` as written doesn't seem to work with `gutentags_cache_dir`, since the file list command doesn't produce file paths relative to the cache directory. I'll look into how to solve this.

My next few days are busy but I'll try to make the requested changes by the weekend. No rush, enjoy your vacation and we'll chat when you're ready. Thanks!

**Comment by smkent** [comment]
I made the requested changes to `gutentags_file_list_command`'s accepted values. Let me know if you'd prefer to use a flat `Dictionary` instead of the nested `markers` `Dictionary`.

I believe this now also works with `gutentags_cache_dir`. When updating tags files with a custom file list command and a cache directory defined, the tag update scripts will prepend each line of output from the custom file list command with the project path, turning each relative path into an absolute path.

I debated about how to pass the (potentially modified) file list command output to `ctags`, `cscope`, etc., and I decided on using a temp file (`$TAGS_FILE.temp` in the same directory as `$TAGS_FILE` and `$TAGS_FILE.lock`), which is then passed as input to the tags executable (e.g. `ctags -L $TAGS_FILE.files [...]`). The *nix scripts could be changed to pass this output to the tags executable with a pipe instead of a temp file, but I'm not sure how to accomplish that in Windows batch files.

As before, I'm not able to fully test on Windows or make sure that the cscope output file works properly. Let me know if you get a chance to test either of those, and I would be happy to fix any issues.

**Comment by ludovicchabant** [comment]
Thanks for the changes! I like having the nested `markers` dictionary because it gives us some room to add a `fallback` or `default` command next to it, along with other things in the future.

Using a `.files` temp file is how I would have done it too I think.

Making this work on Windows was a bit tricky because of how stupid and convoluted `cmd.exe` can get sometimes, but I believe I got it working (see my fixes on top of your changes). If there are any bugs introduced by this hopefully they will be reported.

Thanks!

**Comment by smkent** [comment]
I'm glad everything works now. I have next to no experience writing Windows batch scripts, but it's great to see you got them working without too much trouble.

This makes Gutentags so much easier for me to use. Thanks for merging!
